dropshot/src/server.rs

```diff
@@ -578,15 +615,78 @@ impl<C: ServerContext> HttpServer<C> {
     }
 
     /**
-     * Signals the currently running server to stop and waits for it to exit.
+     * Return the result of registering the server's DTrace USDT probes.
+     *
+     * See [`ProbeRegistration`] for details.
      */
-    pub async fn close(mut self) -> Result<(), String> {
-        self.close_channel
+    pub fn probe_registration(&self) -> &ProbeRegistration {
+        &self.probe_registration
+    }
+}
+
+impl<C: ServerContext, Closer> HttpServer<C, Closer, Joinable> {
+    /// Separates the [HttpServer] from the [Joinable] object which is
+    /// capable of joining it.
+    pub fn get_joiner(self) -> (HttpServer<C, Closer, Unjoinable>, Joinable) {
+        let HttpServer {
+            probe_registration,
+            app_state,
+            local_addr,
+            closer,
+            joiner,
+        } = self;
+
+        (
+            HttpServer {
+                probe_registration,
+                app_state,
+                local_addr,
+                closer,
+                joiner: Unjoinable {},
+            },
+            joiner,
+        )
+    }
+}
+
+impl<C: ServerContext, Joiner> HttpServer<C, Closeable, Joiner> {
```

Comment:
Probably a silly question, but: why can't we just impl `close()` directly on the `HttpServer` in this block? (i.e., why have the impl of `close()` be on the `Closeable` instead of just using that as a marker?)

Reply:
I actually *did* impl `close()` on the `HttpServer`, but it wasn't nearby in the code. Re-located.